src/StoreApi/Routes/Checkout.php

```diff
@@ -141,105 +183,80 @@ protected function get_route_update_response( \WP_REST_Request $request ) {
 	/**
 	 * Update and process an order.
 	 *
-	 * 1. Obtain Draft Order
+	 * 1. Obtain Draft Order.
 	 * 2. Process Request
 	 * 3. Process Customer
 	 * 4. Validate Order
 	 * 5. Process Payment
 	 *
 	 * @throws RouteException On error.
-	 * @param \WP_REST_Request $request Request object.
-	 * @return \WP_REST_Response
+	 * @param WP_REST_Request $request Request object.
+	 * @return WP_REST_Response
 	 */
-	protected function get_route_post_response( \WP_REST_Request $request ) {
-		// --
-		// 1. Obtain Draft Order
-
-		// Update customer first since orders will be created using that data.
+	protected function get_route_post_response( WP_REST_Request $request ) {
+		/**
+		 * Obtain Draft Order and process request data.
+		 *
+		 * Note: Customer data is persisted from the request first so that OrderController::update_addresses_from_cart
+		 * uses the up to date customer address.
+		 */
 		$this->update_customer_from_request( $request );
-
-		// Get the order, including validation and reserving stock.
-		$order = $this->create_or_update_draft_order();
-
-		// --
-		// 2. Process Request
-
-		// If any form fields were posted, update the order (e.g. order note, payment method).
-		$this->update_order_from_request( $order, $request );
+		$this->create_or_update_draft_order();
+		$this->update_order_from_request( $request );
 
 		// Validate payment method before proceeding.
-		if ( $order->needs_payment() ) {
-			// Will throw and return 400 if payment method missing or invalid.
-			$this->get_request_payment_method_id( $request );
+		if ( $this->order->needs_payment() && empty( $request['payment_method'] ) ) {
+			throw new RouteException(
+				'woocommerce_rest_checkout_missing_payment_method',
+				__( 'No payment method provided.', 'woo-gutenberg-products-block' ),
+				400
+			);
```

Comment:
`Checkout::update_order_from_request` already checks to see if `$request['payment_method']` exists. I think we could move this validation inside there to keep the logic flow in this method cleaner? Also it would help surface this error earlier in the stack because `update_order_from_request` already does some validation around whether a payment method exists for the payment method id passed in the request.

Reply:
That will need tweaking because it's required on POST but not on UPDATE right now. Ill see if that difference is still needed.